### PR TITLE
Add FK constraints on verification codes

### PR DESCRIPTION
**Pull request by sethvargo**
**Release Note**

```release-note
NONE
```

### CODE REVIEW

File: pkg/database/migrations.go [hunk 1]
```diff
@@ -2452,6 +2452,37 @@ func (db *Database) Migrations(ctx context.Context) []*gormigrate.Migration {
 				)
 			},
 		},
+		{
+			ID: "00115-AddVerificationCodeConstraints",
+			Migrate: func(tx *gorm.DB) error {
+				return multiExec(tx, `
+					ALTER SEQUENCE verification_codes_id_seq AS BIGINT;
```
Comment (mikehelmick) [inline pkg/database/migrations.go:2459]: any concerns w/ these taking a long time to run?

Comment (mikehelmick) [reply pkg/database/migrations.go:2459]: this could a long table lock over the verification_codes table.

File: pkg/database/vercode.go [hunk 2]
```diff
@@ -89,29 +91,47 @@ type VerificationCode struct {
...
-	IssuingExternalID string `gorm:"column:issuing_external_id; type:varchar(255);"`
+	IssuingExternalID string `gorm:"column:issuing_external_id; type:text;"`
 }
 
 // BeforeSave is used by callbacks.
 func (v *VerificationCode) BeforeSave(tx *gorm.DB) error {
+	v.IssuingUserIDPtr = uintPtr(v.IssuingUserID)
+	v.IssuingAppIDPtr = uintPtr(v.IssuingAppID)
+
+	if v.RealmID == 0 && v.Realm == nil {
```
Comment (mikehelmick) [inline pkg/database/vercode.go:115]: would this be better to just

if v.Realm == nil {
  error
}
v.RealmID = v.Realm.ID

File: pkg/database/vercode.go [hunk 3]
```diff
@@ -89,29 +91,47 @@ type VerificationCode struct {
...
-		return fmt.Errorf("validation failed: %s", strings.Join(msgs, ", "))
+	return v.ErrorOrNil()
+}
+
+// AfterFind runs after the verification code is found.
+func (v *VerificationCode) AfterFind(tx *gorm.DB) error {
+	v.IssuingUserID = uintValue(v.IssuingUserIDPtr)
+	v.IssuingAppID = uintValue(v.IssuingAppIDPtr)
+
+	if v.Realm == nil {
```
Comment (mikehelmick) [inline pkg/database/vercode.go:131]: do we really need to load the realm for every verification code read?

File: pkg/database/vercode.go [hunk 4]
```diff
@@ -69,12 +68,15 @@ type VerificationCode struct {
...
 
-	RealmID       uint   // VerificationCodes belong to exactly one realm when issued.
-	Code          string `gorm:"type:varchar(512)"`
-	LongCode      string `gorm:"type:varchar(512)"`
-	UUID          string `gorm:"type:uuid;default:null"`
-	Claimed       bool   `gorm:"default:false"`
-	TestType      string `gorm:"type:varchar(20)"`
+	// RealmID is the realm in which this verification code was issued. Codes belong to exactly one realm.
+	RealmID uint `gorm:"column:realm_id; type:integer; not null;"`
+	Realm   *Realm
```
Comment (mikehelmick) [inline pkg/database/vercode.go:73]: I'm not sure the loading of the realm relation is 100% necessary here...

File: pkg/database/vercode.go [hunk 5]
```diff
@@ -204,7 +224,8 @@ func (r *Realm) FindVerificationCode(db *Database, code string) (*VerificationCo
 
 	var vc VerificationCode
 	if err := db.db.
-		Where("realm_id = ? AND (code IN (?) OR long_code IN (?))", r.ID, hmacedCodes, hmacedCodes).
+		Joins("JOIN realms ON realms.id = verification_codes.realm_id").
```
Comment (mikehelmick) [inline pkg/database/vercode.go:227]: per the earlier comment - not loading the realm relation here can avoid a join.

File: pkg/database/vercode.go (hunk 2 above)
Comment (sethvargo) [reply pkg/database/vercode.go:115]: Some places set `Realm` and some places set `RealmID`. We really only care if `RealmID` is set during saving, but this catches both. It's also what we do in MobileApp and a few other models already.

Comment (sethvargo) [reply pkg/database/vercode.go]: But doesnt matter because removed realm

File: pkg/database/vercode.go (hunk 4 above)
Comment (sethvargo) [reply pkg/database/vercode.go]: Done

File: pkg/database/vercode.go (hunk 3 above)
Comment (sethvargo) [reply pkg/database/vercode.go]: Done

File: pkg/database/vercode.go (hunk 5 above)
Comment (sethvargo) [reply pkg/database/vercode.go]: Done

File: pkg/database/migrations.go (hunk 1 above)
Comment (mikehelmick) [reply pkg/database/migrations.go:2459]: discussed & benchmarked offline - all good.

**Review by mikehelmick — APPROVED** [review]

File: pkg/database/migrations.go (hunk 1 above)
Comment (sethvargo) [reply pkg/database/migrations.go:2459]: Upgrading column types is constant (int -> bigint and varchar -> text), so I'm not concerned there.

Adding the FK constraints will lock for a full table scan for each foreign key. I ran this migration on local dev against a seed run with stats (which created 120k verification codes) and it took ~0.15s.

**Pull request merged**